src/Framework/AspNetCoreAnalyzers/test/RouteHandlers/DisallowNonParsableComplexTypesOnParametersTest.cs

```diff
@@ -0,0 +1,487 @@
+
+public class Customer : IParsable<Customer>
+{
+    public static Customer Parse(string s, IFormatProvider provider)
+    {
+        if (TryParse(s, provider, out Customer customer))
+        {
+            return customer;
+        }
+        else
+        {
+            throw new ArgumentException(s);
+        }
+    }
+
+    public static bool TryParse(string s, IFormatProvider provider, out Customer result)
+    {
+        result = new Customer();
+        return true;
+    }
+}
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Route_Parameter_withParsableComplexType_viaMethodConvention_Works()
+    {
+        // Arrange
+        var source = $$"""
+using System;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapPost("/customers/{customer}/contacts", (Customer customer) => {});
+
+public class Customer
+{
+    public static bool TryParse(string s, IFormatProvider provider, out Customer result)
+    {
+        result = new Customer();
+        return true;
+    }
+}
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Route_Parameter_withHttpContextBindableComplexType_viaImplicitIBindableFromHttp_Works()
+    {
+        // Arrange
+        var source = $$"""
+using System;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+
+var webApp = WebApplication.Create();
+webApp.MapGet("/customers/{customer}", (Customer customer) => {});
+
+public class Customer : IBindableFromHttpContext<Customer>
+{
+    public static async ValueTask<Customer?> BindAsync(HttpContext context, ParameterInfo parameter)
+    {
+        return new Customer();
+    }
+}
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Route_Parameter_withNullableType_Works()
+    {
+        // Arrange
+        var source = $$"""
+using System;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapGet("/customers/{customer}/contacts", (int? customer) => {});
+""";
+
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source);
+    }
+
+    [Fact]
+    public async Task Handler_Parameter_withUnbindableComplexType_Fails()
+    {
+        // Arrange
+        var source = $$"""
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Builder;
+
+var webApp = WebApplication.Create();
+webApp.MapPost(
+    "/customers",
+    ({|#0:[FromBody]Customer customer|}) => {
+    });
+
+public class Customer
+{
+}
+""";
+
+        var expectedDiagnostic = new DiagnosticResult(DiagnosticDescriptors.RouteHandlerParamterComplexTypeIsNotBindable)
+            .WithArguments("customer", "Customer")
+            .WithLocation(0);
+        // Act
+        await VerifyCS.VerifyAnalyzerAsync(source, expectedDiagnostic);
```

Comment:
This should work. The `source` being tested with a POCO `Customer` is perfectly valid.

```csharp
webApp.MapPost(
    "/customers",
    ([FromBody] Customer customer) => {
    });
```

`[FromBody] Customer customer` and just `Customer customer` have the same behavior in this case assuming there's no `Customer` registered as a service. It's deserialized as JSON. Uploading `{}` as the body works just fine.

Even if `Customer` did implement `BindAsync`, `[FromBody]` is enough to force JSON deserialization (unless the parameter type is `Stream` or `PipeReader`). We check for the `[FromBody]` attribute here. We could rename `BindParameterFromBody` to `BindParameterFromJsonBody` in this PR to clarify that code a bit.

https://github.com/dotnet/aspnetcore/blob/600eb9aa53c052ec7327e2399744215dbe493a89/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L699-L713

And only fall back to BindAsync later here:

https://github.com/dotnet/aspnetcore/blob/600eb9aa53c052ec7327e2399744215dbe493a89/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L798-L801

I'm just linking to the RDF code for reference. I wouldn't expect anyone to figure this out just from reading the code.  However, it would be good to run these sample apps to verify it really is broken. It's also probably be a good idea to update the docs at https://github.com/dotnet/AspNetCore.Docs/blob/96d01d63c1886691c96fafa4a299f01948a8ff34/aspnetcore/fundamentals/minimal-apis/includes/parameter-binding.md#binding-precedence to clarify that `[FromBody]` means JSON deserialization, Stream or PipeReader and not a parameter type with a `BindAsync` method which may or may not read from the body.

That said, and this isn't the case for this test method, it might make sense to use `BindAsync` if it exists even when there is a `[FromBody]` attribute. We don't do that today, but I bet people would expect us to, and I doubt anyone is relying on us *not* calling `BindAsync` in that case.

Reply:
> `[FromBody] Customer customer` and just `Customer customer` have the same behavior in this case assuming there's no `Customer` registered as a service. It's deserialized as JSON. Uploading `{}` as the body works just fine.

Note quite true from my experimentation:

```(Customer customer) => {}``` will use ```BindAsync(...)``` if it is available and fallback to JSON deserialize if it isn't.
```([FromBody]Customer customer) => {}``` will always JSON deserialize.

Reply:
You're correct. I was focusing on the case where there is no `BindAsync` or `TryParse` implementation like in this test, but I didn't make that clear.